### PR TITLE
CDH: add en/decrypt support for eHSM-KMS

### PR DESCRIPTION
**Pull request by 1570005763**
Add eHSM-KMS client to support en/decrypt operation.

eHSM-KMS related tasks mentioned in #353 will be further developed based on it.

### CODE REVIEW

**Comment by fitzthum** [comment]
Cool, I will check this out once we are done with the release.

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    use crate::{plugins::ehsm::client::EhsmKmsClient, Decrypter, Encrypter};
+
+    #[ignore]
+    #[tokio::test]
+    async fn test_create_key() {
+        let key_spec = "EH_AES_GCM_256";
+        let provider_settings = json!({
+            "app_id": "86f0e9fe-7f05-4110-9f65-a224ddee1233",
+            "endpoint": "https://172.16.1.1:9002",
```
Comment (fitzthum) [inline confidential-data-hub/kms/src/plugins/ehsm/client.rs:165]: Is this a real instance running somewhere? Is it guaranteed to be running?

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 2]
```diff
@@ -316,9 +316,10 @@ mod tests {
 
     use crate::{plugins::aliyun::client::AliyunKmsClient, Decrypter, Encrypter};
 
-    #[ignore]
     #[rstest]
+    #[ignore]
     #[case(b"this is a test plaintext")]
+    #[ignore]
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client.rs:322]: The change upon this file seems not related to the PR.

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                base_url: endpoint.to_owned(),
+                appid: app_id.to_owned(),
+                apikey: api_key.to_owned(),
+            }
+        })
+    }
+
+    /// build client with parameters that have been exported to environment.
+    pub fn new_from_env() -> Result<Self> {
+
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/ehsm/client.rs:45]:
```suggestion

```

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    /// to access kms is saved.
+    pub async fn from_provider_settings(provider_settings: &ProviderSettings) -> Result<Self> {
+        let provider_settings: EhsmProviderSettings =
+            serde_json::from_value(Value::Object(provider_settings.clone())).map_err(|e| {
+                Error::EhsmKmsError(format!("parse provider setting failed: {e}"))
+            })?;
+
+        let credential_path = format!(
+            "{EHSM_IN_GUEST_DEFAULT_KEY_PATH}/credential_{}.json",
+            provider_settings.app_id
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/ehsm/client.rs:62]: About the way to read credential. Currently we follow the steps:
1. CDH get credential from the KBS
2. CDH put the credential under `/run/..`
3. CDH reads the credential under `/run/..`

@fitzthum once mentioned that this will expose the plaintext of credentials to the guest and suggest that the credentials be kept in the memory of the CDH. I think it is a good way, and the steps could be
1. CDH get credential from the KBS
2. CDH sets its own process's env the credential with specific keys
3. Concrete KMS plugin reads the credential from the env

Which I'd like to put another PR and let's talk more about that then.

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    #[ignore]
+    #[case(b"this is a test plaintext")]
+    #[ignore]
+    #[case(b"this is a another test plaintext")]
+    #[tokio::test]
+    async fn key_lifetime(#[case] plaintext: &[u8]) {
+        let key_spec = "EH_AES_GCM_256";
+        let provider_settings = json!({
+            "app_id": "86f0e9fe-7f05-4110-9f65-a224ddee1233",
+            "endpoint": "https://172.16.1.1:9002",
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/ehsm/client.rs:190]: same as mentioned

File: confidential-data-hub/kms/src/plugins/mod.rs [hunk 6]
```diff
@@ -14,6 +14,9 @@ pub mod aliyun;
 
 pub mod kbs;
 
+#[cfg(feature = "ehsm")]
+pub mod ehsm;
+
 #[derive(AsRefStr, EnumString)]
 pub enum DecryptorProvider {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/mod.rs:21]: We might also need to add `eHSM-KMS` to `new_decryptor`

**Review by Xynnn007 — COMMENTED** [review]
Thanks @1570005763 pretty good code! Some nits

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs (hunk 1 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/ehsm/client.rs:165]: Sure! it is an eHSM-KMS service that I have deployed on my local machine. And all tests pass successfully.
AA31A454-D5A6-463D-AC06-3864653EF76D

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs (hunk 2 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs:322]: Indeed, it was my mistake. I will open another PR about this change.

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs (hunk 4 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/ehsm/client.rs:61]: Glad to hear that. eHSM-KMS itself supports reading the credential from environment variables, which will make the transition easy to implement.

**Comment by 1570005763** [comment]
Modifications to the Aliyun KMS client have been removed.

File: confidential-data-hub/kms/src/plugins/mod.rs (hunk 6 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/mod.rs:21]: Will do.

File: confidential-data-hub/kms/Cargo.toml [hunk 7]
```diff
@@ -13,6 +13,7 @@ bincode = { workspace = true, optional = true }
 chrono = { workspace = true, optional = true }
 const_format.workspace = true
 crypto = { path = "../../attestation-agent/deps/crypto", optional = true }
+ehsm_client = {git = "https://github.com/intel/ehsm", optional = true }
```
Comment (Xynnn007) [inline confidential-data-hub/kms/Cargo.toml:16]: It is good to have a `rev` here to fix the version of the upstream version, which is helpful to avoid breaking caused by api changes

Comment (1570005763) [reply confidential-data-hub/kms/Cargo.toml]: That's helpful. Thanks!

Comment (Xynnn007) [reply confidential-data-hub/kms/Cargo.toml]: : )  Could you help to squash the commits into one?

Comment (1570005763) [reply confidential-data-hub/kms/Cargo.toml]: Sure thing. : )

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs (hunk 1 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/ehsm/client.rs:164]: By the way, it's hard to run these tests on CoCo CI, since eHSM-KMS service relies on SGX. But I add a readme file about how to set up eHSM-KMS service locally. So that anyone can test eHSM-KMS Client on a device with SGX capabilities.

**Review by Xynnn007 — APPROVED** [review]
lgtm.

File: confidential-data-hub/kms/src/plugins/ehsm/client.rs (hunk 1 above)
Comment (fitzthum) [reply confidential-data-hub/kms/src/plugins/ehsm/client.rs:164]: Ok. I guess that's why these tests are marked as ignore. You might want to replace your IP with a phony one.

**Comment by 1570005763** [comment]
@fitzthum I have fixed these issues, thanks for your review!
